GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs

```diff
@@ -0,0 +1,466 @@
+                return false;
+            }
+            else
+            {
+                message = null;
+            }
+
+            message = null;
+            return true;
+        }
+
+        public bool TryQueryNewestVersion(out Version newVersion, out string message)
+        {
+            try
+            {
+                IList<IPackageSearchMetadata> queryResults = this.nuGetFeed.QueryFeedAsync(this.nugetUpgraderConfig.PackageFeedName).GetAwaiter().GetResult();
+
+                // Find the latest package
+                IPackageSearchMetadata highestVersion = null;
+                foreach (IPackageSearchMetadata result in queryResults)
+                {
+                    if (highestVersion == null || result.Identity.Version > highestVersion.Identity.Version)
+                    {
+                        highestVersion = result;
+                    }
+                }
+
+                if (highestVersion != null &&
+                    highestVersion.Identity.Version.Version > this.installedVersion)
+                {
+                    this.latestVersion = highestVersion;
+                }
+
+                newVersion = this.latestVersion?.Identity?.Version?.Version;
+
+                if (newVersion != null)
+                {
+                    this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - new version available: installedVersion: {this.installedVersion}, latestAvailableVersion: {highestVersion}");
+                    message = $"New version {highestVersion.Identity.Version} is available.";
+                    return true;
+                }
+                else if (highestVersion != null)
+                {
+                    this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - up-to-date");
+                    message = $"Latest available version is {highestVersion.Identity.Version}, you are up-to-date";
+                    return true;
+                }
+                else
+                {
+                    this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - no versions available from feed.");
+                    message = $"No versions available via feed.";
+                }
+            }
+            catch (Exception ex)
+            {
+                this.tracer.RelatedError($"{nameof(this.TryQueryNewestVersion)} failed with: {ex.Message}");
+                message = ex.Message;
+                newVersion = null;
+            }
+
+            return false;
+        }
+
+        public bool TryDownloadNewestVersion(out string errorMessage)
+        {
+            if (this.latestVersion == null)
+            {
+                errorMessage = "No new version to download. Query for latest version to ensure a new version is available before downloading.";
+                return false;
+            }
+
+            using (ITracer activity = this.tracer.StartActivity(nameof(this.TryDownloadNewestVersion), EventLevel.Informational))
```

Comment:
This is not currently going to be reported via telemetry, just wanted to confirm if that is intentional.